core/src/price_estimation/clients/generic_client.rs

```diff
@@ -1,79 +1,91 @@
-mod api;
-
-use super::{PriceSource, TokenData};
+use super::super::{PriceSource, TokenData};
 use crate::http::HttpFactory;
 use crate::models::TokenId;
 use anyhow::{anyhow, Context, Result};
-use api::{DexagApi, DexagHttpApi};
 use futures::{
     future::{self, BoxFuture, FutureExt as _},
     lock::Mutex,
 };
 use std::collections::HashMap;
 
-struct ApiTokens {
+/// Provides a generic interface to communicate in a standardized way
+/// with specific API token implementations
+pub trait GenericToken {
+    /// Symbol describing the ERC20 token represented by the type instance
+    fn symbol(&self) -> &str;
+}
+
+#[cfg_attr(test, mockall::automock(type Token=tests::MockGenericToken;))]
+pub trait Api: Sized {
+    type Token: GenericToken + Sync + Send;
+
+    /// Creates a new HTTP interface for connecting to the API
+    fn bind(http_factory: &HttpFactory) -> Result<Self>;
+
+    fn get_token_list<'a>(&'a self) -> BoxFuture<'a, Result<Vec<Self::Token>>>;
+
+    /// Returns the price of one unit of `from` expressed in `to`.
+    /// For example `get_price("ETH", "DAI")` is ~220.
+    fn get_price<'a>(&'a self, from: &Self::Token, to: &Self::Token) -> BoxFuture<'a, Result<f64>>;
+
+    /// Returns a string representing the reference coin in the stead of OWL for this API
+    /// Could be different from "OWL", e.g., when the API does not offer prices with
+    /// respect to OWL
+    fn stable_coin_symbol() -> &'static str;
```

Comment:
One think I considered would be to change the `stable_coin_symbol` to take `&self`. So the reasons I can think of for this would be:
- Makes it so the reference token can be configurable (you could have a Dexag client with DAI as the reference and one with USDC) - this could eventually allow us to configure via command line or environment variable what we want to use for the reference token.
- Makes the mocking and testing a little easier.

Although, its 100% fine how it is.

Reply:
Taking `&self` is reasonable, but the changes needed to implement this proposal would make a similar issue to the one explained in [my other comment](https://github.com/gnosis/dex-services/pull/1040/files#r448462317) arise: The only place where something that is not an abstract type is created is in `price_estimation.rs`, so the call to the constructor would have to look like `DexagClient::new(http_factory, tokens.clone(), &"DAI")`. I don't quite like that because it's not immediate what the role of this "DAI" is.
I see however that if the changes from the linked comment were somehow introduced then this change would definitely be easy to do.

Reply:
We can implement this if it is needed in the future.